java/arcs/core/analysis/BoundedAbstractElement.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.analysis
+
+/**
+ * A class that lifts any type [V] with a special bottom and top value.
+ *
+ * This class may be used to represent the elements of a lattice when implementing the
+ * [AbstractValue] interface. Specifically, it provides helpers for various methods in the
+ * [AbstractValue] interface that have well-known semantics with respect to bottom (the smallest
+ * value) and top (the largest value) in the lattice. For example, `bottom.join(other)` is always
+ * equal to `other` for any value of `other`, because `bottom` is the lowest value in the lattice.
+ *
+ * Here is a sample use of this class for implementing an abstract domain of integer elements,
+ * where the elements are ordered by [Int]'s `<=` operator.
+ * ```
+ * class IntegerDomain(
+ *     val abstractInt: BoundedAbstractElement<Int>
+ * ) : AbstractValue<IntegerDomain> {
+ *     override infix fun join(other: IntegerDomain): IntegerDomain {
+ *         return abstractInt.join(other.abstractInt) { a, b -> maxOf(a, b) }
+ *     }
+ *     // . . .
+ * }
+ * ```
+ */
+data class BoundedAbstractElement<V : Any> private constructor(
```

Comment:
I don't believe you need to specify `<V : Any>`, you can just say `<V>` (here and throughout file)

Reply:
Actually, the `requireNotNull` calls in the implementation does not type check. I get the following error: 
`Not satisfied: inferred type T? is not subtype of Any`

See this [stackoverflow](https://stackoverflow.com/questions/48921447/not-satisfied-inferred-type-t-is-not-subtype-of-any) for some context.

Reply:
Thanks, I didn't realise that would happen. Probably best to leave it the way you have it then.